llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,285 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+  memcpy(strTabData + strTabSize, pName, secNameSize);
+
+  delete[] m_sections[m_strtabSecIdx].data;
+
+  m_sections[m_strtabSecIdx].data = strTabData;
+  m_sections[m_strtabSecIdx].secHead.sh_size = strTabSize + secNameSize;
+}
+
+// =====================================================================================================================
+// Add the reloc related symbols and sections based on the input reloc
+//
+// @param reader : The ElfReader to the input fragment elf bin
+// @param inputReloc : The input reloc
+template <class Elf>
+void ElfWriter<Elf>::addRelocSymbols(const ElfReader<Elf> &reader, typename Elf::Reloc inputReloc) {
+  ElfSymbol relocSym = {};
+  reader.getSymbol(inputReloc.r_symbol, &relocSym);
+
+  // Create a section named origSectionName.cached
+  char *secName = new char[strlen(relocSym.secName) + strlen(CachedRodataSectionSuffix)];
+  strcpy(secName, relocSym.secName);
+  strcat(secName, CachedRodataSectionSuffix);
+
+  for (auto section : m_sections) {
+    if (strcmp(section.name, secName) == 0) {
+      return;
+    }
+  }
+
+  // If the section does not exist, create a new section and update the related symbols
+  SectionBuffer newSection = {};
+  ElfSectionBuffer<Elf64::SectionHeader> *inputRodataSection = nullptr;
+  std::vector<ElfSymbol> inputRodataSymbols;
+  auto inputRodataSecIndex = reader.GetSectionIndex(relocSym.secName);
+  reader.getSectionDataBySectionIndex(inputRodataSecIndex, &inputRodataSection);
+  reader.GetSymbolsBySectionIndex(inputRodataSecIndex, inputRodataSymbols);
+
+  createNewSection(secName, inputRodataSection, &newSection);
+
+  // Update the data from the input section
+  auto secData = new uint8_t[inputRodataSection->secHead.sh_size];
+  memcpy(secData, inputRodataSection->data, inputRodataSection->secHead.sh_size);
+  newSection.data = secData;
+  newSection.secHead.sh_size = inputRodataSection->secHead.sh_size;
+
+  m_sections.push_back(newSection);
+
+  // Update symbols
+  for (auto rodataSymbol : inputRodataSymbols) {
+    ElfSymbol symbol = {};
+    // Create a new symbol named origSymbolName_cached
+    char *newSymName = new char[strlen(rodataSymbol.pSymName) + strlen(CachedRodataSymbolSuffix)];
+    strcpy(newSymName, rodataSymbol.pSymName);
+    strcat(newSymName, CachedRodataSymbolSuffix);
+    symbol.secIdx = m_sections.size() - 1;
+    symbol.secName = secName;
+    symbol.size = rodataSymbol.size;
+    symbol.pSymName = newSymName;
+    symbol.value = rodataSymbol.value;
+    symbol.info.all = rodataSymbol.info.all;
+    symbol.nameOffset = InvalidValue;
+    m_symbols.push_back(symbol);
+  }
+}
+
+// =====================================================================================================================
+// Get the reloc symbol index based on the input reloc symbol
+//
+// @param inputRelocSymbol : The input reloc symbol
+template <class Elf> uint32_t ElfWriter<Elf>::getRelocSymbolIndex(ElfSymbol inputRelocSymbol) {
```

Comment:
I think taking just the symbol name as an argument is sufficient and doesn’t need to copy the whole `ElfSymbol` struct.

Reply:
Addressed.